test/test_images/latencymako/main.go

```diff
@@ -0,0 +1,466 @@
+	ctx, q, qclose, err := mako.Setup(ctx)
+	if err != nil {
+		log.Fatalf("Failed to setup mako: %v", err)
+	}
+
+	// Use a fresh context here so that our RPC to terminate the sidecar
+	// isn't subject to our timeout (or we won't shut it down when we time out)
+	defer qclose(context.Background())
+
+	// Wrap fatalf in a helper or our sidecar will live forever.
+	fatalf = func(f string, args ...interface{}) {
+		qclose(context.Background())
+		log.Fatalf(f, args...)
+	}
+
+	printf("Mako configured")
+
+	// --- Warmup phase
+
+	if warmupSeconds > 0 {
+		warmup(warmupSeconds)
+	} else {
+		printf("Warmup skipped")
+	}
+
+	printf("--- BENCHMARK ---")
+	printf("Configuring channels")
+
+	// --- Allocate channels
+
+	// We need those estimates to allocate memory before benchmark starts
+	var estimatedNumberOfMessagesInsideAChannel uint64
+	var estimatedNumberOfTotalMessages uint64
+
+	for _, pacer := range pacerSpecs {
+		totalMessages := uint64(pacer.rps * int(pacer.duration.Seconds()))
+		// Add a bit more, just to be sure that we don't allocate
+		totalMessages = totalMessages + uint64(float64(totalMessages)*0.1)
+		// Queueing theory stuff: Given our channels can process 50 rps, queueLength = arrival rps / 50 rps = pacer.rps / 50
+		queueLength := uint64(pacer.rps / 50)
+		if queueLength < 10 {
+			queueLength = 10
+		}
+		estimatedNumberOfTotalMessages = estimatedNumberOfTotalMessages + totalMessages
+		if queueLength > estimatedNumberOfMessagesInsideAChannel {
+			estimatedNumberOfMessagesInsideAChannel = queueLength
+		}
+	}
+
+	printf("Estimated channel size: %v", estimatedNumberOfMessagesInsideAChannel)
+	printf("Estimated total messages: %v", estimatedNumberOfTotalMessages)
+
+	// Create all channels
+	sentCh = make(chan sentState, estimatedNumberOfMessagesInsideAChannel)
+	deliveredCh = make(chan deliveredState, estimatedNumberOfMessagesInsideAChannel)
+	receivedCh = make(chan receivedState, estimatedNumberOfMessagesInsideAChannel)
+	resultCh = make(chan eventStatus, estimatedNumberOfMessagesInsideAChannel)
+
+	// Small note: both receivedCh and resultCh depends on receive thpt and not send thpt
+	// but we don't care since this is a pessimistic estimate and receive thpt < send thpt
+
+	printf("Starting CloudEvents receiver")
+
+	// Start the events receiver
+	startCloudEventsReceiver(processReceiveEvent)
+
+	printf("Starting latency processor")
+
+	// Start the goroutine that will process the latencies and publish the data points to mako
+	go processLatencies(q, estimatedNumberOfTotalMessages)
+
+	if warmupSeconds > 0 {
+		// Just let the channel relax a bit
+		time.Sleep(5 * time.Second)
+	} else {
+		// sleep 30 seconds before sending the events
+		// TODO(Fredy-Z): this is a bit hacky, as ideally, we need to wait for the Trigger/Subscription that uses it as a
+		//                Subscriber to become ready before sending the events, but we don't have a way to coordinate between them.
+		time.Sleep(30 * time.Second)
```

Comment:
I think this isn't necessary anymore, but it doesn't hurt for now.

Reply:
I think it's wise to keep it for now, it could be useful maybe if we want to chain more resources between sink and source